### ISSUE TITLE
Error Code Registry

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#749]
In GitLab by @plinss on Jun 3, 2020, 10:35

I'm in the process of developing a few extensions for flake8, it would be useful if there was a central registry of all existing error codes so that I can ensure my extensions don't conflict with anyone else's. I've already seen examples of extensions changing their codes due to conflicts.

It doesn't have to be anything sophisticated, a simple markdown file in a repo that extension developers can submit MRs to would be fine.

In principle, it should at least list:

* All the built-in error codes
* Prefixes reserved for flake8 (and maybe any carved out areas where extensions would be welcome to extend)
* The error code prefixes reserved for each extension (a link to the extension's home page would be great)
* And ideally a full list of the error codes provided by each extension along with a brief description (and possibly a link to the extension's documentation of that code).

The full list will also help users who are trying to figure out what error codes referenced in config files or `noqa` comments actually do. Currently these can be very hard to find for random extensions...

Of course I don't expect the flake8 developers to maintain the entire list (just the built-in codes), extension developers and end-users can submit their own entries and should be encouraged to do so. A link to the registry and a paragraph explaining its use in the docs for developing extensions should do.

FWIW, I'm willing to submit a MR to the flake8 docs and an initial pass at a registry file if this idea is supported.

**Comment by asottile** [comment]
In GitLab by @asottile on Jun 3, 2020, 11:02

I think it would be fine to document a simple list that is:

```markdown
- [plugin-name](https://example.com/link-to-plugin) - PFX
```

for the builtin ones this would look like:

```markdown
- [pyflakes](https://github.com/pycqa/pyflakes) - F
- [pycodestyle](https://github.com/pycqa/pycodestyle) - E, W
- [mccabe](https://github.com/pycqa/mccabe) - C
```

I kinda don't want to get into individual codes or descriptions of them -- that should be up to the individual tools' documentation to go into details there (plus if we were to duplicate their docs here it would always be out of date and invite many noisy PRs, I don't really have interest in reviewing many hundreds of PRs to update grammar / synchronize codes with someone else's tool)

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jun 3, 2020, 11:40

I think just having what you describe will lead to hundreds of PRs which is
why I've rejected this in the past. There are at least 2 closed duplicates
if I remember correctly

**Comment by asottile** [comment]
In GitLab by @plinss on Jun 3, 2020, 11:53

@asottile a simpler list works for me. My one suggestion would be to link to the extension's (and built-in) documentation of their error codes, if there is one (maybe on the error code prefix?). e.g.

```
- [pyflakes](https://github.com/pycqa/pyflakes) - [F](https://flake8.pycqa.org/en/latest/user/error-codes.html)
```

@sigmavirus24 1) Sorry for not finding the duplicate issues, I did search. 2) I'm sympathetic to not wanting to create hundreds of PRs. But the fact that this has come up before only helps to underscore that there's a real need.

Does @asottile's suggestion of a prefix only list satisfy your concern? We're looking at 1 PR per extenson, I think that's a reasonable burden for the given value. Also, if we use a dedicated repo to host the list, it's entirely possible to add outside collaborators to help review and manage the PRs. I volunteer myself to help out with that.

And as I said, I'm also happy to create the initial list, a contributing guide, and a MR to update the flake8 docs.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jun 4, 2020, 12:44

> But the fact that this has come up before only helps to underscore that there's a real need.

I would be wary to conflate need and desire.

> Also, if we use a dedicated repo to host the list, it's entirely possible to add outside collaborators to help review and manage the PRs. I volunteer myself to help out with that.

Then this doesn't need to live in PyCQA or flake8 if it doesn't need to be in our docs and anyone can and should create it at any time without needing @asottile or myself invovled.

That said, there's way too many things you're not accounting for:

- What happens when two plugins already have the same prefix? Which one gets listed? Do both get listed? Does the first to register get listed? Does the one with more installations get listed?

- What links are acceptable? GitHub READMEs are (in my experience) unstable places for documentation. Do you require readthedocs.io links or github.io pages links?

- What is a prefix? I can think of one plugin that uses `I9xx` and leaves the rest of the `I[0-8]xx` space empty and open for anyone to use.

In short, a text-based registry here seems like a nightmare to manage, especially as soon as conflicts arise and confusion around which prefixes are which.

I understand the _desire_ to avoid stepping on other people's toes, but I don't think it's the most valuable use of @asottile's or my time.

**Comment by asottile** [comment]
In GitLab by @plinss on Jun 4, 2020, 15:15

> Then this doesn't need to live in PyCQA or flake8 if it doesn't need to be in our docs and anyone can and should create it at any time without needing @asottile or myself invovled.

Agreed, but there's value in a single list being blessed as the "official" list, having multiple lists maintained in different places would be worse and just lead to more [confusion](https://xkcd.com/927/).

I don't care where it lives, I just thought having a repo under your org would make it easier to find and more obvious that this one is the one to pay attention to. I'm happy to setup a completely independent repo but it would make sense for it to be referenced in the flake8 extension developer docs and considered "official". I'm not interested in getting into a war with two or three other registries.

If you're ok with that, I'll spin one up and submit a PR to flake8 for the docs to refer to it. I'll use a generic name but would like to use the flake8 name in it, something like `flake8-extension-registry` (suggestions welcome). If we go that route, do you prefer it to be on gitlab.com, github.com, or mirrored on both with one being the primary? Personally I prefer GitLab but I accept that GitHub has a bigger audience at the moment.

> That said, there's way too many things you're not accounting for:

> * What happens when two plugins already have the same prefix? Which one gets listed? Do both get listed? Does the first to register get listed? Does the one with more installations get listed?

For existing extensions that have conflicts, I don't think having one in the list first would give it any special status, or having a popularity contest would be beneficial to anyone. But finding these kinds of situations is a *good* thing, and encouraging extension developers to work out which one will change is a benefit to the community.

I've already seen it happen, so there is an issue here worth addressing IMO.

Part of the registry contributing docs should be a conflict resolution process, e.g. if you find someone listed a code you're already using, here's how you work it out with them... we can also list both until it gets sorted (and if they're all unmaintained or people get stubborn, it may never get sorted, but that's OK, at least the fact that there's a conflict will be obvious for anyone considering installing them and maybe encourage someone else to fork one and fix it.)

The registry can just reflect reality and encourage good behavior.

> * What links are acceptable? GitHub READMEs are (in my experience) unstable places for documentation. Do you require readthedocs.io links or github.io pages links?

I'm not sure it's worth being link police, or bless a documentation host, or add extra requirements to extension developers. Encourage people to use stable links that actually refer to extension home pages and/or code docs where available. Extensions will die, links will go stale. Anyone who notices a problem can submit an issue or PR. 

Not having a link is fine too, but they should be encouraged, an entry without links still has value. If someone finds a broken link, or if a domain gets taken over by an ad wall, whoever finds it can file an issue or PR to fix it or drop the link. PRs for links to spam or unrelated sites will simply be rejected by the maintainers. 

* What is a prefix? I can think of one plugin that uses `I9xx` and leaves the rest of the `I[0-8]xx` space empty and open for anyone to use.

That's fine, it should be whatever prefix(es) the plugin registers for its entry point, no? A registry would make it easier and safer for developers to carve out and share numeric spaces inside a text prefix, which is a good thing. Better guidance about selecting prefixes in the flake8 docs would be good too, a registry will help make it easier. 

> In short, a text-based registry here seems like a nightmare to manage, especially as soon as conflicts arise and confusion around which prefixes are which.

Yes, there will be work. But frankly a wild-west approach where you never know if a new extension you install is going to step on another one is a bigger nightmare. Or if you're developing an extension and you have no idea if the prefix you want to use has been taken, and google searches for 'X' aren't useful. Is every developer really expected to find every other extension?

Not having a central registry just spreads the pain over a larger audience which makes it less noticeable for those not experiencing it directly. Not good for the community, tho, and can discourage participation from developers and create frustrated users.

> I understand the *desire* to avoid stepping on other people's toes, but I don't think it's the most valuable use of @asottile's or my time.

And I'm not expecting or asking either of you to spend your time on it. As I said, I'm happy to spin it up, and I'm happy to review PRs and collaborate with anyone who wants to volunteer to maintain the list. If either of you *wants* to participate, great, but no need to, the community is large enough and there are people with good will and time to offer.

I'm happy to take a quick pass through extensions that I know about and are listed in a few popular lists to kick the registry off, but policing the internet to find extensions shouldn't be anyone's job. A link to the registry in the flake8 docs with a note telling developers to pick an unused prefix and submit a PR to add it should suffice. Word will get around for existing extensions.

All I'm asking from you guys is your blessing and your willingness to accept a PR to update the flake8 docs referring to the registry (to your standards of course). If you want to spin up a repo in your org and give me access (or just review my PRs to it), that'd be great too, but is not required for this to work.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Jun 5, 2020, 15:35

> Agreed, but there's value in a single list being blessed as the "official" list, having multiple lists maintained in different places would be worse and just lead to more [confusion](https://xkcd.com/927/).

In the _years_ since the first issue talking about an error code registry, no one has started one. Not because people are forbidden but because no one needs it enough. If there were enough need, someone would have done it and that would be the religiously proper one.

I can't imagine someone seeing a list or registry maintained by someone and saying "Let me make my own". Even if there is a semi-official version, anyone who would otherwise have been motivated to make their own, will still make their own. It's the nature of open source - anyone can do what they want if they have the ability and free time and nothing can stop them.

It sounds like you just want to beat anyone over the head with the "official"-ness of having it in the PyCQA and that's not what the organization is designed for. It's designed as a place for people to collaborate and reduce bus factor, not as any real "authority" on anything. Nothing we do is official in any sense of the phrase. It's just convenient.

> If you're ok with that, I'll spin one up and submit a PR to flake8 for the docs to refer to it. I'll use a generic name but would like to use the flake8 name in it, something like `flake8-extension-registry` (suggestions welcome). If we go that route, do you prefer it to be on gitlab.com, github.com, or mirrored on both with one being the primary? Personally I prefer GitLab but I accept that GitHub has a bigger audience at the moment.

If you start a registry, absolutely, we'll reference it in the docs. I don't care where you host the code. We aren't really picky. Again, there's no real authority beyond the word being in the name.

> And I'm not expecting or asking either of you to spend your time on it.

It's been my experience that folks wanting to start something in an org don't stick around and leave the "owners" of the org (the people with permission enough to close issues and merge contributions) with the bag. It's why the PyCQA is willing to accept new projects if they're popular enough, but we don't guarantee maintenance. If we start this in the PyCQA and it doesn't take off, I'll be harassed enough on social media to likely just start merging anything because I'm a believer in backwards compatibility and a stable web. That's exactly why I _don't_ want to be involved.

If at some point y'all have a reasonable maintenance team and it's a popular project, we can re-evaluate since it might be easier to have the structure of an organization + team to help make sure enough people are "admins".

**Comment by asottile** [comment]
In GitLab by @plinss on Jun 5, 2020, 20:05

> In the *years* since the first issue talking about an error code registry, no one has started one. Not because people are forbidden but because no one needs it enough. If there were enough need, someone would have done it and that would be the religiously proper one.

Also possible, those who asked for it weren't willing to put in the work and just published their plugin hoping for the best and moved on (or saw pushback and decided it wasn't worth the effort).

> I can't imagine someone seeing a list or registry maintained by someone and saying "Let me make my own". Even if there is a semi-official version, anyone who would otherwise have been motivated to make their own, will still make their own. It's the nature of open source - anyone can do what they want if they have the ability and free time and nothing can stop them.

My concern wasn't so much someone seeing a registry and deciding to make their own, but more of starting one because they didn't know there already was one. And then you have diverging sets... Of course anyone is free to start another one or fork the first one if they feel they can do better. I'm no stranger to open source, having been shipping it since the 90's.

> It sounds like you just want to beat anyone over the head with the "official"-ness of having it in the PyCQA and that's not what the organization is designed for. 

Absolutely not. By "official" I really only meant "mentioned in the flake8 docs" so people know where to look and don't have to trip over it by accident or know to go searching for it. I've been working in web standards for a long time so I think to look for registries where there's potential for conflicts. I imagine most casual plugin developers wouldn't otherwise think about it much.

> It's designed as a place for people to collaborate and reduce bus factor, not as any real "authority" on anything. Nothing we do is official in any sense of the phrase. It's just convenient.

Of course. 

My thinking about the registry is that it should be a reference available for plugin developers and maybe users, nothing more. I never imagined the registry as have any authority in itself, e.g. people are of course free to ignore it and stomp all over other plugins. I just want to create a system where they aren't doing that by accident. I have no intention of policing anything.

> If you start a registry, absolutely, we'll reference it in the docs.

Perfect. I'll be submitting a MR to the docs shortly. I have a few other plates in the air at the moment but obviously this is something I care about.

> It's been my experience that folks wanting to start something in an org don't stick around and leave the "owners" of the org (the people with permission enough to close issues and merge contributions) with the bag. It's why the PyCQA is willing to accept new projects if they're popular enough, but we don't guarantee maintenance. If we start this in the PyCQA and it doesn't take off, I'll be harassed enough on social media to likely just start merging anything because I'm a believer in backwards compatibility and a stable web. That's exactly why I *don't* want to be involved.

Totally fair.

> If at some point y'all have a reasonable maintenance team and it's a popular project, we can re-evaluate since it might be easier to have the structure of an organization + team to help make sure enough people are "admins".

Sounds good.

**Comment by asottile** [comment]
In GitLab by @plinss on Jun 5, 2020, 20:05

closed

**Comment by asottile** [comment]
In GitLab by @thejcannon on Aug 26, 2020, 08:14

An example of how a error code resource would possibly be useful is in [`flakehell`](https://flakehell.readthedocs.io/)'s `codes` support (where it does its best to tell you which codes a plugin uses).

`flakehell` has a bunch of methods, each trying to extract the codes from the plugins one way or another: `https://github.com/life4/flakehell/blob/master/flakehell/_logic/_extractors.py`. I'd love for this to be simple.

How would the owners feel about a standard _mechanism_ for plugins to list codes/descriptions? Then it would be easy for a registry to scrape all the plugins :tada:, it would be easy for something like `flakehell` to scrape codes in a standard way, and `flake8` itself need not support anything other than the mechanism.

(I'm happy to open a separate issue if y'all are open to it, since it would be a separate issue)

Edit for clarification:
The _mechanism_ allude to is essentially a way of _strongly encouraging_ plugin authors enumerate their codes and descriptions in their plugin. This is nice to the humans and the machines.
